service/chain/follower.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package chain
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/onflow/flow-go/follower"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/operation"
+	"github.com/rs/zerolog"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type Follower struct {
+	log zerolog.Logger
+
+	consensus follower.ConsensusFollower
+	db        *badger.DB
+
+	availableHeight uint64
+	height          uint64
+	blockID         flow.Identifier
+}
+
+func FromFollower(log zerolog.Logger, follower follower.ConsensusFollower, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		consensus:       follower,
+		height:          math.MaxUint64,
+		availableHeight: math.MaxUint64,
+		blockID:         flow.ZeroID,
+	}
+
+	follower.AddOnBlockFinalizedConsumer(f.OnBlockFinalized)
+
+	return &f
+}
+
+func (f *Follower) OnBlockFinalized(finalizedBlockID flow.Identifier) {
+	var height uint64
+	err := f.db.View(operation.RetrieveFinalizedHeight(&height))
+	if err != nil {
+		f.log.Error().Err(err).Msg("Could not retrieve finalized block height")
+		return
+	}
+
+	f.availableHeight = height
+}
+
+func (f *Follower) Root() (uint64, error) {
+	var height uint64
+	err := f.db.View(operation.RetrieveRootHeight(&height))
+	if err != nil {
+		return 0, fmt.Errorf("could not look up root height: %w", err)
+	}
+	return height, nil
+}
+
+func (f *Follower) Commit(height uint64) (flow.StateCommitment, error) {
+
+	blockID, err := f.block(height)
+	if errors.Is(err, storage.ErrNotFound) {
+		return flow.StateCommitment{}, dps.ErrFinished
+	}
+	if errors.Is(err, dps.ErrTimeout) {
+		return flow.StateCommitment{}, dps.ErrTimeout
+	}
+	if err != nil {
+		return flow.StateCommitment{}, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	var commit flow.StateCommitment
+	err = f.db.View(operation.LookupStateCommitment(blockID, &commit))
+	if err != nil {
+		return flow.StateCommitment{}, fmt.Errorf("could not look up commit: %w", err)
+	}
```

Comment:
This can fail for a finalized block. We can only index blocks that are finalized _AND_ executed. In order to determine execution status, we have to check if there is a `flow.ExecutionResult` for the block yet. Rather than waiting for the `flow.Seal` in a block payload and getting the commit from that, using the execution result will allow us to index/serve the block before it is sealed, which will reduce latency significantly.

Reply:
We will have latency anyway for anything that requires the block to be executed, won't we? I assumed that we would have to always only index data from finalized and executed blocks, which is why the `block` method which advances to the next block, only does so if it is within the executed height. Indexing partial data seems like it would add a lot of complexity.

Reply:
If we don't wait for a seal, how will we know that an execution result we have is actually correct?

Reply:
Okay, I see your other comment below might explain this. It makes sense, but doesn't this mean that this would only be a temporary optimization? Eventually, we will need to wait for a seal right?

Reply:
@smnzhu @awfm9 For now I'm going with just using the latest finalized height as the height we can index to simplify things.

Reply:
@smnzhu The DPS indexes blocks that are both _finalized_ and _executed_. Unless there is a malicious or faulty execution node, this is sufficient to know the result.

As we get the results directly from the execution node, we don't need to wait for sealing to serve this information, which drastically reduces latency, considering how slow the sealing process is, and how much slower it will become in the future. This is in line with what we discussed with @AlexHentschel.

When we have seals in a finalized block, we make sure to sanity check the state commitment from the seal against the state commitment from the execution result. If that fails, we just crash for now, as it means that an execution node produced an incorrect execution result.